### ISSUE TITLE
It would be nice to have an API to verify that the context is internally consistent

### ISSUE THREAD

**Issue by xnox** [openssl/openssl#14607]
Whilst configuring a context one can set all the things on it, in various order.

- min prototocol versions
- max protocol versions
- enable that suite/chipher/checksums
- disable given suite/chipher/checksums
- change security level
- load certs (in case of client auth, or server cert)
- change security callback function

Once one is done, one may by accident or intentionally configure an invalid context. I.e. max protocol version lower than min one, those that are rejected by the security hook, with chiphers that are not at all defined for given protocols, with certs using rejected checksums, and key sizes that are way too small, etc.

It would be nice, to have a _verify_consistent() api that would check some/most of the above things to ensure that when one thinks they have configured a valid context, it actually is internally consistent. Instead of finding that out the hardway, later on, when trying to establish connection and realizing that all the things fail - irrespective of what the other side of the connection offers.

**Comment by tiran** [comment]
This feature would be very helpful for Python's ssl module.

**Comment by nhorman** [comment]
@mattcaswell @t8m is this feasible?

**Comment by t8m** [comment]
IMO some of these inconsistencies are already checked and impossible. However there definitely are things like the SECLEVEL vs. the certificate used where we could do more checks. Given there are probably many such possible checks this would be a non-trivial effort. I worry that although such api would definitely be useful it is not really worth the effort needed.

**Comment by paulidale** [comment]
Would checking that the ciphersuite list is non-empty be a reasonable start?

I too worry that this might be a bottomless pit.

**Comment by xnox** [comment]
> Would checking that the ciphersuite list is non-empty be a reasonable start?
> 

It would help a lot to check that / return such status upon doing changes to the context.

I.e. if setting min/max version results in no  ciphersuite that is useful to know. Then the user of the API can figure out if they are setting more things later on, or if it is unexpected that after applying all the settings there is nothing left.

> I too worry that this might be a bottomless pit.

Doing 80% of the way will help a lot.
